shotover/src/transforms/kafka/sink_cluster/connections.rs

```diff
@@ -159,6 +179,74 @@ impl Connections {
             ConnectionState::Unopened
         }
     }
+
+    /// Open a new connection to the requested Destination.
+    /// If the connection is succesfully created, the old connection is replace with the new one.
+    /// Otherwise the old connection is just deleted with no replacement and an error is returned.
+    pub async fn handle_connection_error(
+        &mut self,
+        connection_factory: &ConnectionFactory,
+        authorize_scram_over_mtls: &Option<AuthorizeScramOverMtls>,
+        sasl_mechanism: &Option<String>,
+        nodes: &[KafkaNode],
+        destination: Destination,
+        error: anyhow::Error,
+    ) -> Result<()> {
+        let address = match destination {
+            Destination::Id(id) => {
+                &nodes
+                    .iter()
+                    .find(|x| x.broker_id == id)
+                    .unwrap()
+                    .kafka_address
+            }
+            Destination::ControlConnection => self.control_connection_address.as_ref().unwrap(),
+        };
+
+        let connection = connection_factory
+            .create_connection(address, authorize_scram_over_mtls, sasl_mechanism)
+            .await
+            .context("Failed to create a new connection");
+
+        match connection {
+            Ok(connection) => {
+                // Recreating the node succeeded.
+                // So store it as the new connection, as long as we werent waiting on any responses in the old connection
+                let connection = KafkaConnection::new(
+                    authorize_scram_over_mtls,
+                    sasl_mechanism,
+                    connection,
+                    None,
+                )?;
+                let old = self.connections.insert(destination, connection);
+
+                if old.map(|old| old.pending_requests_count()).unwrap_or(0) > 0 {
+                    Err(error.context("Succesfully reopened outgoing connection but previous outgoing connection had pending requests."))
```

Comment:
good point.
I fixed the logic to first delete the old connection from the hashmap and then only reinsert the new connection if we succeed.
I previously avoided doing it this way for performance reasons, but since this is an error path that level of performance concern is not needed here.